### PR TITLE
Get PDFs to jump to their pages again

### LINKED ISSUE #1245: Jump to the right page
> Please provide us with the following information:
> ---------------------------------------------------------------
Hello,

Following the recent update, upon clicking the PDF file reference link, the file displayed started from the very first file page instead of the referenced one as worked previously. Anyone observing the same thing?

Thanks!

### This issue is for a: (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ x ] regression (a behavior that used to work and stopped in a new release)
```

Linux

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

### PR DESCRIPTION
**Pull request by pamelafox**
## Purpose

Fixes #1245

This PR has a few fixes, but the biggest one is a fix for regression introduced by #1160 which replaced citation URLs with blob URLs. My fix remembers the hash and adds it to the blob URL after, which seems to work correctly in Edge. I couldn't find any documentation as to whether blob URLs formally support hash fragments, but they seem to!

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Click on citations in answers that have #page= hash, see PDF viewer jump to that page
* Switch to GPT4V env, click on citations with no hash, see them open fine

### CODE REVIEW

File: app/backend/approaches/chatapproach.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ class ChatApproach(Approach, ABC):
     Make sure the last question ends with ">>".
     """
 
-    query_prompt_template = """Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge.
+    query_prompt_template = """Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base.
```
Comment (pamelafox) [inline app/backend/approaches/chatapproach.py:40]: This was caught by a reader of my blog post today-https://blog.pamelafox.org/2024/02/rag-techniques-cleaning-user-questions.html

Doubt it affects anything as GPTs dont mind lil errors.

File: app/frontend/src/components/AnalysisPanel/ThoughtProcess.tsx [hunk 2]
```diff
@@ -12,9 +12,9 @@ interface Props {
 export const ThoughtProcess = ({ thoughts }: Props) => {
     return (
         <ul className={styles.tList}>
-            {thoughts.map(t => {
+            {thoughts.map((t, ind) => {
                 return (
-                    <li className={styles.tListItem}>
+                    <li className={styles.tListItem} key={ind}>
```
Comment (pamelafox) [inline app/frontend/src/components/AnalysisPanel/ThoughtProcess.tsx:17]: We were getting a bunch of React errors locally from missing key= values. An index isnt the best thing to use but it should be okay for our situation.

File: app/frontend/src/components/SupportingContent/SupportingContent.tsx [hunk 3]
```diff
@@ -6,32 +6,23 @@ interface Props {
...
-    content: string;
-}
-
 export const SupportingContent = ({ supportingContent }: Props) => {
     const textItems = Array.isArray(supportingContent) ? supportingContent : supportingContent.text;
     const imageItems = !Array.isArray(supportingContent) ? supportingContent?.images : [];
     return (
         <ul className={styles.supportingContentNavList}>
-            {textItems.map(c => {
+            {textItems.map((c, ind) => {
```
Comment (pamelafox) [inline app/frontend/src/components/SupportingContent/SupportingContent.tsx:14]: Ditto, this was just fixing index. And I moved the HTML inside the class as I couldn't deal with writing the correct Typescript for the separate component. :)

File: app/frontend/src/components/AnalysisPanel/AnalysisPanel.tsx [hunk 4]
```diff
@@ -34,12 +34,19 @@ export const AnalysisPanel = ({ answer, activeTab, activeCitation, citationHeigh
     const fetchCitation = async () => {
         const token = client ? await getToken(client) : undefined;
         if (activeCitation) {
+            // Get hash from the URL as it may contain #page=N
```
Comment (mattgotteiner) [inline app/frontend/src/components/AnalysisPanel/AnalysisPanel.tsx:37]: This makes sense since we download the full object url. Appreciate your fix

**Review by mattgotteiner — APPROVED** [review]

**Pull request merged**